File: java/patterns-use-cases/src/main/java/my/example/queue/AsyncTaskService.java
package my.example.queue;

import dev.restate.sdk.Context;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.Service;
import dev.restate.sdk.http.vertx.RestateHttpEndpointBuilder;

@Service
public class AsyncTaskService {

    public record TaskOpts(String key, String taskName, String payload) {}

    @Handler
    public String runTask(Context ctx, TaskOpts params) {
        return someHeavyWork(params);
    }

    private String someHeavyWork(TaskOpts params) {
        return "someHeavyWork";
    }

    public static void main(String[] args) {
         RestateHttpEndpointBuilder.builder().bind(new AsyncTaskService()).buildAndListen();
    }
}

File: kotlin/patterns-use-cases/src/main/kotlin/my/example/queue/AsyncTaskService.kt
package my.example.queue

import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.Service
import dev.restate.sdk.http.vertx.RestateHttpServer
import dev.restate.sdk.kotlin.endpoint.endpoint
import kotlinx.serialization.Serializable

@Service
class AsyncTaskService {
  @Handler
  suspend fun runTask(params: TaskOpts): String {
    return params.someHeavyWork()
  }
}

fun main() {
  RestateHttpServer.listen(endpoint { bind(AsyncTaskService()) })
}

@Serializable class TaskOpts

fun TaskOpts.someHeavyWork(): String {
  return "someHeavyWork"
}
